controllers/usersignup/usersignup_controller.go

```diff
@@ -200,12 +251,97 @@ func (r *Reconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.
+		// The migration process is as follows:
+		//
+		// 1) Check if the UserSignup has been migrated.  This is done by checking if a UserSignup with a name equal
+		//    to the encoded username exists (while also checking that the currently reconciling UserSignup isn't actually it)
+		//
+		// 2) If the UserSignup hasn't been migrated, then migrate it by creating a new UserSignup with the name set to
+		//    the encoded username, and the "migration-replaces" annotation set to the name of this (the old) UserSignup
+		//
+		// 3)
+		//
+		// Other things to remove after migration is completed:
+		//
+		// 1) The setStatusMigrationFailedLookup function in status.go
+		// 2) The setStatusMigrationFailedCreate function in status.go
+		// 3) The setStatusMigrationFailedCleanup function in status.go
+		// 4) The UserSignupUserMigrationFailed constant in status.go
+		// 5) The EncodeUserIdentifier function from this file
+		// 6) The above block of code in this function that deletes the original UserSignup if the "migration-replaces" annotation is set
+		encodedUsername := EncodeUserIdentifier(userSignup.Spec.Username)
+		if userSignup.Name != encodedUsername {
+			migratedUserSignup := &toolchainv1alpha1.UserSignup{}
+			err := r.Client.Get(context.TODO(), types.NamespacedName{
+				Namespace: request.Namespace,
+				Name:      encodedUsername,
+			}, migratedUserSignup)
+			if err != nil {
+				if errors.IsNotFound(err) {
+					// Create the migrated UserSignup here.
+					//
+					// We need to:
+					//
+					// 1) Copy the existing UserSignup
+					// 2) Override the new UserSignup resource name with the encoded username
+					// 3) Set the starting status to Deactivated (technically we could let the reconciler function do this
+					//    when it reconciles the migrated UserSignup, but it shouldn't hurt to set this status up front)
+					migratedUserSignup = userSignup.DeepCopy()
+					migratedUserSignup.Name = encodedUsername
+					r.setStatusDeactivated(migratedUserSignup, "")
+					err = r.Client.Create(ctx, migratedUserSignup)
+					if err != nil {
+						// If there was an error creating the migrated UserSignup, then set the status and requeue
+						return reconcile.Result{}, r.wrapErrorWithStatusUpdate(logger, userSignup,
+							r.setStatusMigrationFailedCreate, err, "Failed to migrate UserSignup")
+					}
+				}
+				// Error reading the object - requeue the request.
+				return reconcile.Result{}, r.wrapErrorWithStatusUpdate(logger, userSignup,
+					r.setStatusMigrationFailedLookup, err, "Failed to lookup migrated UserSignup")
+			}
+		}
+
 		return reconcile.Result{}, r.updateStatus(logger, userSignup, r.setStatusDeactivated)
 	}
 
 	return reconcile.Result{}, r.ensureNewMurIfApproved(logger, config, userSignup)
 }
 
+// EncodeUserIdentifier is a temporary function used by the migration procedure
+// REMOVE AFTER MIGRATION IS COMPLETE
+func EncodeUserIdentifier(subject string) string {
```

Comment:
We already have this function somewhere, right? We use it when creating a usersignup I suppose... Is it just a copy of that function (in case you don't want to move it from reg-service to common for a temporal migration step or something like this)?

Reply:
We do have it in registration service, I was tempted to move it to toolchain-common however after the migration is complete the only place it will still be used will only be registration service again which is why I simply made a temporary copy of it.